adr-tools 2.1.0 (new formula)

kevinjalbert opened the pull request:
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Add new formula for adr-tools (v2.1.0) - a command-line tools for working with Architecture Decision Records.

bfontaine commented on Formula/adr-tools.rb:
```diff
@@ -0,0 +1,21 @@
+    # Override adr-config with custom config to point to homebrew dirs
+    system "echo '#!/bin/bash' > src/adr-config"
+    system "echo 'echo adr_bin_dir=\"#{bin}\"' >> src/adr-config"
+    system "echo 'echo adr_template_dir=\"#{prefix}\"' >> src/adr-config"
```
Use `Pathname#write` instead:
```ruby
(buildpath/"src/adr-config").write <<-EOS.undent
  your content
  goes here
EOS
```

Please ask upstream how we could specify this config without overriding their file.

kevinjalbert replied:
As I'm trying to overwrite the `src/adr-config` file, using `Pathname#write` I end up with an error:

```
Error: Will not overwrite /private/tmp/adr-tools-20170501-58028-1kqrmh5/adr-tools-2.1.0/src/adr-config
```

Is there a suggested way to override an existing file? Or at least end up with different content for an existing file? Looking through the [Formula-Cookbook](https://github.com/Homebrew/brew/blob/master/docs/Formula-Cookbook.md) I see [inreplace](https://github.com/Homebrew/brew/blob/master/docs/Formula-Cookbook.md#inreplace). I think that might be a better solution here?

-----

For:
> Please ask upstream how we could specify this config without overriding their file.

There was [an attempt](https://github.com/Homebrew/homebrew-core/pull/3947) to put a previous version of this tool in homebrew in which a similar discussion took place. [Upstream work was done resulting in](https://github.com/npryce/adr-tools/blob/master/doc/adr/0007-invoke-adr-config-executable-to-get-configuration.md#consequences):

> Packagers can write their own implementation of adr-config that outputs configuration that matches the platform's installation conventions, and deploy it next to the adr script.

bfontaine replied:
Thanks! Yes, `inreplace` is a better solution.

bfontaine commented on Formula/adr-tools.rb:
```diff
@@ -0,0 +1,21 @@
+    system "echo 'echo adr_template_dir=\"#{prefix}\"' >> src/adr-config"
+
+    prefix.install Dir["src/*.md"] # Places the base templates in `prefix`
+    bin.install Dir["src/*"]       # Copies the rest to `prefix/bin`
```
Please remove these comments.

bfontaine commented on Formula/adr-tools.rb:
```diff
@@ -0,0 +1,21 @@
+
+  test do
+    assert_match(/Run 'adr help COMMAND' for help on a specific command./,
+                 pipe_output("#{bin}/adr help"))
```
Please write a test that does something more specific than printing the help. Also, use `shell_output` instead of `pipe_output` if you don’t pipe anything.

kevinjalbert commented:
@bfontaine I've addressed the code review comments.

One thing I'm not too sold on is the use of `inreplace` to fit the shebang in.

Another alternative that I could do is:

```ruby
    File.open("src/adr-config", "w") do |file|
      file.write <<-EOS.undent
        #!/bin/bash
        echo 'adr_bin_dir=\"#{bin}\"'
        echo 'adr_template_dir=\"#{prefix}\"'
      EOS
    end
```

Although, I'm not too sure if that is any better from a formula perspective.

bfontaine commented:
You could remove the default file then write one again, e.g.: 
```ruby
config = buildpath/"src/adr-config"
# Add a comment here explaining why we’re removing the default file.
config.unlink
config.write <<-EOS.undent
  # ...
EOS
```

kevinjalbert commented:
@bfontaine nice! I didn't know about the `unlink`. Not 100% sure if my comment is sufficient and I'm open to suggestions if needed.
